### PR TITLE
Add Play Map button to map editor

### PR DESCRIPTION
**Pull request by PunkPun**

### CODE REVIEW

**Review by Mailaender — DISMISSED** [review]
Works as promised.

File: OpenRA.Mods.Common/Widgets/Logic/MissionBrowserLogic.cs [hunk 1]
```diff
@@ -181,7 +191,7 @@ protected override void Dispose(bool disposing)
 
 		void CreateMissionGroup(string title, IEnumerable<MapPreview> previews, Action onExit)
 		{
-			var header = ScrollItemWidget.Setup(headerTemplate, () => true, () => { });
+			var header = ScrollItemWidget.Setup(headerTemplate, () => false, () => { });
```
Comment (dragunoff) [inline OpenRA.Mods.Common/Widgets/Logic/MissionBrowserLogic.cs]: This is a hack to make the headers have a different appearance:
image
Without it they look wrong:
image

I guess the proper fix here would be to change the background of the ScrollItem in the header template. But in any case it's not related to what this PR is doing so please revert that line.

**Review by dragunoff — CHANGES_REQUESTED** [review]
Do we need to pick last modified map in the lobby and mission browser if we are adding the "Play Map" button in the editor? Is there a use case for that? :thinking: I think the play map/back to editor buttons cover the mapper's needs (to quickly test the map they're editing).

File: OpenRA.Mods.Common/Widgets/Logic/MissionBrowserLogic.cs (hunk 1 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Widgets/Logic/MissionBrowserLogic.cs]: It is related, as when you click the play button the Map in mission selector needs to be preselected. I'll adjust the visuals

Comment (dragunoff) [reply OpenRA.Mods.Common/Widgets/Logic/MissionBrowserLogic.cs]: I see. But then again - do we want to preselect the last edited mission? This doesn't seem like the right UX to me.

Comment (PunkPun) [reply OpenRA.Mods.Common/Widgets/Logic/MissionBrowserLogic.cs]: If you edit / upload a mission you are very likely going to want to play it. I've applied the same reasoning as I did with MapChooser

**Comment by PunkPun** [comment]
Depends on #20200

**Comment by dragunoff** [comment]
@PunkPun Does the "Play map" button (the last commit here) depend on anything in #20200? Can we split this and review/merge independently?

**Comment by PunkPun** [comment]
yes it depends on #20200 adding the ability to preselect maps in mission selector (so no we can't split)

**Comment by abcdefg30** [comment]
> Can we split this and review/merge independently?

Fwiw, I think it would be good to split an extra "go back to the editor"-button PR from this and focus just on editor->ingame here.

**Comment by PunkPun** [comment]
> Fwiw, I think it would be good to split an extra "go back to the editor"-button PR from this and focus just on editor->ingame here.

I've separated them into commits. I don't think it's worth splitting as this is already a fairly small and self-contained PR. I've already split of 2 PR's from this

**Comment by AspectInteractive2** [comment]
If I go to the "Play" lobby of a map but do not proceed and instead click exit, then try to go back into it again, as soon as I click "Menu" I get the below error. I was able to replicate the bug consistently.

```
Exception of type `System.InvalidOperationException`: The current order generator may not be changed from synced code
   at OpenRA.Sync.AssertUnsynced(String message) in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Game\Sync.cs:line 210
   at OpenRA.World.set_OrderGenerator(IOrderGenerator value) in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Game\World.cs:line 157
   at OpenRA.World.CancelInputMode() in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Game\World.cs:line 167
   at OpenRA.Mods.Common.Widgets.Logic.MenuButtonsChromeLogic.OpenMenuPanel(MenuButtonWidget button, WidgetArgs widgetArgs) in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Mods.Common\Widgets\Logic\Ingame\MenuButtonsChromeLogic.cs:line 93
   at OpenRA.Mods.Common.Widgets.Logic.MenuButtonsChromeLogic.<>c__DisplayClass5_2.<.ctor>b__1() in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Mods.Common\Widgets\Logic\Ingame\MenuButtonsChromeLogic.cs:line 46
   at OpenRA.Mods.Common.Widgets.ButtonWidget.<.ctor>b__41_5(MouseInput _) in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Mods.Common\Widgets\ButtonWidget.cs:line 82
   at OpenRA.Mods.Common.Widgets.ButtonWidget.HandleMouseInput(MouseInput mi) in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Mods.Common\Widgets\ButtonWidget.cs:line 201
   at OpenRA.Widgets.Widget.HandleMouseInputOuter(MouseInput mi) in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Game\Widgets\Widget.cs:line 420
   at OpenRA.Widgets.Ui.HandleInput(MouseInput mi) in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Game\Widgets\Widget.cs:line 100
   at OpenRA.DefaultInputHandler.<>c__DisplayClass5_0.<OnMouseInput>b__0() in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Game\Input\InputHandler.cs:line 50
   at OpenRA.DefaultInputHandler.OnMouseInput(MouseInput input) in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Game\Input\InputHandler.cs:line 51
   at OpenRA.Platforms.Default.Sdl2Input.PumpInput(Sdl2PlatformWindow device, IInputHandler inputHandler, Nullable`1 lockedMousePosition) in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Platforms.Default\Sdl2Input.cs:line 149
   at OpenRA.Platforms.Default.Sdl2PlatformWindow.PumpInput(IInputHandler inputHandler) in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Platforms.Default\Sdl2PlatformWindow.cs:line 473
   at OpenRA.Renderer.EndFrame(IInputHandler inputHandler) in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Game\Renderer.cs:line 311
   at OpenRA.Game.RenderTick() in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Game\Game.cs:line 733
   at OpenRA.Game.Loop() in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Game\Game.cs:line 841
   at OpenRA.Game.Run() in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Game\Game.cs:line 873
   at OpenRA.Game.InitializeAndRun(String[] args) in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Game\Game.cs:line 308
   at OpenRA.Launcher.Program.Main(String[] args) in C:\DEV\OpenRA_PlayButton\OpenRA\OpenRA.Launcher\Program.cs:line 32
```

**Comment by PunkPun** [comment]
Yes, this is the reason why this PR is in draft. Nobody knew how to fix that

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs [hunk 2]
```diff
@@ -133,6 +154,8 @@ public class IngameMenuLogic : ChromeLogic
 		bool leaving;
 		bool hideMenu;
 
+		static bool lastGameEditor = false;
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs:157]: This static flag on the widget seems hacky, i'm not sure I have a better suggeston but it feels like we should be managing this at something higher level that keeps track of global UI/Game state and not within the widget directly.

Comment (PunkPun) [reply OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameMenuLogic.cs:157]: I believe we had long discussions about how to implement this. This was the best one we came up with

**Review by RoosterDragon — APPROVED** [review]

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ff12f9dfaa4224bb6038a376afeb044728980196...9b7e78370d7f63f4b0702bd661c82d4fd5fb386f)

**Pull request merged**